### ISSUE TITLE
torch.nn.functional.scaled_dot_product_attention is_causal fails for kv-cache case (sequential and further parallel attention)

### ISSUE THREAD

**Issue by JamesGlare** [pytorch/pytorch#144858]
### 🚀 The feature, motivation and pitch

**Behaviour found for torch version 2.2.2**

It would be great if scaled_dot_product_attention could be (easily) used for the case of sequential token generation when a kv-cache is present. However, currently when is_causal is set and a single query vector is put in, the function only compares against the earliest k and v resulting in repeatedly producing the same vector in sequential token generation.

More generally, in cases where further parallel attention is required - even when already a kv-cache has been generated - I found the correct attention matrix difficult to generate. The code I converged to is
`mask = torch.tril(torch.ones(w, w, dtype=torch.bool))[-h:, :]`
with 
w = sequence length of KV-cache
h = sequence length of queries
since a lower-triangular attention mask is required which is all-true for the case of a single query vector.

Is this indeed the intended way to use scaled_dot_product_attention or am I doing something dumb?

### Alternatives

for is_causal=True, I propose to use the attention mask generated by the code above (plus some unsqueezing for broadcasting).

**Comment by vulkomilev** [comment]
I am on it.

**Comment by Chris-hughes10** [comment]
I experienced the same issue and this fix worked for me

**Comment by TimFelixBeyer** [comment]
I think this could be useful too. An even simpler fix could be changing the behavior from 
```python3
temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
```
in the [pseudocode](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) to 
```python3
temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=S-L)
```
I'm interested to hear what the maintainers think @colesbury !

**Comment by Majdoddin** [comment]
Pytorch has now [`causal_lower_right(h, w)`](https://docs.pytorch.org/docs/stable/generated/torch.nn.attention.bias.causal_lower_right.html)

```
  import torch
  import torch.nn.functional as F
  from torch.nn.attention.bias import causal_lower_right

  # KV cache scenario: 1 new query token, 10 cached key/value tokens
  h = 1  # query sequence length
  w = 10 # key/value sequence length (cache)

  batch, num_heads, head_dim = 2, 8, 64
  q = torch.randn(batch, num_heads, h, head_dim)
  k = torch.randn(batch, num_heads, w, head_dim)
  v = torch.randn(batch, num_heads, w, head_dim)

  # Use causal_lower_right for proper KV cache masking
  attn_bias = causal_lower_right(h, w)
  output = F.scaled_dot_product_attention(q, k, v, attn_bias)
```

**Comment by drisspg** [comment]
Closing, see @Majdoddin's comment
